Upload submission functions using a pandas dataframe

benschreyer opened the pull request:
Not sure if this fits with the NumerAPI design pattern/thesis but this would be useful to me in some form in the API for signals and main weekly submission scripts.

Instead of taking a filename argument these overloaded functions take a submission pandas dataframe and upload it in the same way the file is uploaded using a BytesIO file like RAM object

benschreyer committed:
Attempt at submitting pandas dataframe as a signal instead of file

This should write the frame to a RAM binary object and then upload that, would be useful for my submission code
@bensch on rocket chat

benschreyer committed:
Update numerapi.py

benschreyer committed:
Correct commenting

benschreyer committed:
Comments

uuazed commented on numerapi/numerapi.py:
```diff
@@ -6,10 +6,17 @@
 import requests
 import pytz
+import pandas as pd
+
```
what are those newlines for?

uuazed commented on numerapi/numerapi.py:
```diff
@@ -777,6 +784,80 @@ def upload_predictions(self, file_path: str, tournament: int = 8,
         submission_id = create['data']['create_submission']['id']
         return submission_id
 
+    def upload_predictions(self, df: pd.DataFrame, tournament: int = 8,
```
overloading with a different signature doesn't really work in python. Have you tried your code? I am sure only one of the two options (path vs dataframe) works.

uuazed commented on numerapi/numerapi.py:
```diff
@@ -777,6 +784,80 @@ def upload_predictions(self, file_path: str, tournament: int = 8,
+                }
+            }
+            '''
+        arguments = {'filename': os.path.basename(file_path),
```
'file_path' is not defined anywhere. This won't work

uuazed commented on numerapi/numerapi.py:
```diff
@@ -777,6 +784,80 @@ def upload_predictions(self, file_path: str, tournament: int = 8,
+
+        df.to_csv(buffer_csv, index = False)
+
+        with buffer_csv as fh:
```
You copy-pasted lots of code, leading to lots of code duplication. However, the two `upload_predictions' methods just differ in about 3 lines. Much cleaner options would be:
1. add an additional `df` parameter to the existing `upload_predictions` function. In case the function is called with `df` set, do the BytesIO thingy.
or 2. add some `upload_predictions_from_dataframe` function that does the bytesio thing and calls the existing function afterwards.

uuazed commented on numerapi/numerapi.py:
```diff
@@ -6,10 +6,17 @@
 # Third Party
 import requests
 import pytz
+import pandas as pd
```
pandas is a new dependency. It needs to be added in `setup.py` and in `requirements.txt`

uuazed reviewed (changes requested):
Nice addition and a good feature to have!

However, there are plenty of things to improve before this is ready to be merged. There is way too much code duplication, tests are missing and I don't believe your code is working at all right now

benschreyer committed:
Fixing signals and main pandas DataFrame uploads

Removed use of function overloading in favor of adding a new optional argument df to allow uploading a dataframe submission

I have tested  and I was able to upload a dataframe for signals, I will need to wait until round 263 to test for the main tournament since all my model slots are in use

Added new test and main requirements to setup.py, requirements.txt and requirements_tests.txt

added upload call and assertion for testing dataframe upload to tests for signalsapi and numerapi, not really sure how testing works (I get a key error on the response json, not sure how the mock network stuff works yet) I am going to revise that part later

uuazed commented on numerapi/numerapi.py:
```diff
@@ -732,6 +735,15 @@ def upload_predictions(self, file_path: str, tournament: int = 8,
+        buffer_csv = None
+
+        if not df is None:
+            file_path = "predictions.csv"
```
1. not df is None => df is not None. That's the idiomatic way of writing this in Python and is easier to read
2. `file_path` is an argument of the function with a default parameter. No need to set it again here.

uuazed commented on numerapi/numerapi.py:
```diff
@@ -753,9 +765,9 @@ def upload_predictions(self, file_path: str, tournament: int = 8,
+        with open(file_path, 'rb') if df is None else buffer_csv as fh:
             requests.put(
-                submission_auth['url'], data=fh.read(), headers=headers)
+                submission_auth['url'], data=fh.read() if df is None else fh.getvalue(), headers=headers)
```
Add a `buffer_csv.seek(0)` right after writing the csv to it. Then you can just call `read` on it and this line of code doesn't need the `if`

uuazed reviewed (changes requested):
Already looks a lot better!

benschreyer committed:
seek buffer to 0 so .read() works, not is => is not, repetitive arg setting fix

benschreyer committed:
seperated testing for dataframe upload and csv upload

benschreyer committed:
examples in documentation comments
